### DISCUSSION TITLE
Backporting newer features to previous versions

### DISCUSSION THREAD

**Discussion by a60915** [Sitecore/jss#1747] [Help]
### Description

Hello, I was wondering what the general feeling is for backporting newer features to previous versions. I have seen that some features from v21 have been brought to v20. I am interested in contributing to this endeavor but wanted to gauge interest in accepting those types of PRs. I would prefer if it could be part of the main library, but if it can't be, then I would create it as it's own library. 

I admit, I have a selfish interest in this, we are an older version of Sitecore and won't be able to upgrade for a while, so it would be nice to have access to these features today instead of a couple of years from now.

**Comment by pzi** [comment]
We are in a similar boat re: not in a position to upgrade for a little bit... I was wondering if you had specific features in mind?

**Reply by mjarosch** [reply]
@pzi Ideally, Pull Requests to back port features would be accepted to past versions. Hence my starting of this discussion to gauge support. Plan B would be to distribute this as an add-on library. Not sure how far back is feasible to support, but open to ideas.

I started working on a [POC](https://github.com/mjarosch/jss-backport-test) of this over the weekend.I am supporting a 10.2 instance, so adding support for Robots.txt, Redirects and Multisite using searches instead of sites is feasible.

I am running into issues where Typescript classes aren't "extendable", so back porting additions to those may take some creativity or may mean renamed replacements instead of enhancements in some cases.

**Reply by mjarosch** [reply]
@pzi Are there any particular features that you would be interested in?

**Reply by pzi** [reply]
Hey @mjarosch, I put together a rough wishlist with things that would suit us for our purposes :)

Unreleased
- On Demand ISR https://github.com/Sitecore/jss/pull/1674 
- fast-deep-equal https://github.com/Sitecore/jss/pull/1722
- sc_site cookie attr https://github.com/Sitecore/jss/pull/1746

Released
- richtext links https://github.com/Sitecore/jss/pull/1718
- prefetch link option in richtext https://github.com/Sitecore/jss/pull/1517
- model update for item type https://github.com/Sitecore/jss/pull/1219
- any update to Link component
  - https://github.com/Sitecore/jss/pull/1375
  - https://github.com/Sitecore/jss/pull/1226
  - https://github.com/Sitecore/jss/pull/1425
- image component fill prop https://github.com/Sitecore/jss/pull/1336
- removal of getStaticProps/getServerSideProps https://github.com/Sitecore/jss/pull/1753
- next 13/14 support (stretch goal ;) )

And in general, our cyber security is hounding us for updates of Axios :)
- update axios https://github.com/Sitecore/jss/issues/1343

**Reply by mjarosch** [reply]
@pzi how would you feel about a library that replaces the Sitecore jss library? Basically the jss codebase with back ports of graphql queries and other enhancements?

**Reply by pzi** [reply]
While the thought has crossed our minds as well, it would create quite a dependency on another community that's not "sitting within Sitecore" to keep it maintained in somewhat lock-step with this repo. Plus, in case there is an issue I am unsure how (Sitecore) support would look like given it's a 3rd party dep. In other words I'd be hesitant to replace this library but could probably get behind using it in addition to it (even though we'd potentially face similar issues). I'll ask around in our team to hear their thoughts.

**Comment by illiakovalenko** [comment]
@mike-jarosch-aw The primary challenge here is that backporting these features is not a straightforward task. As we move forward, we will also need to maintain support for them and increase QA efforts. I will bring this issue to the attention of our product, and we will provide you with a response soon.
